Refactor Overseer and make object matching more strict

andrew-uoa opened the pull request:
This PR overhauls the object-matching system in the `Overseer` class, to try to make it clearer when identifiers are being used for matching, and when type-specific fields are being used, and to reduce the risk of matching incorrectly, which could lead to e.g. associating datafiles with the wrong dataset.

Broadly, the changes are:

- Overseer changes
  - Remove `get_objects_by_fields()` as it was mostly a wrapper around `_get_objects_from_mytardis()`
  - Replace `get_objects()` with `get_matching_objects()` and `get_objects_by_identifier()`
    - `get_objects()` carried a risk of being used incorrectly, as it used the `object_type` input to get the match key name, but relied on the caller passing in the correct value corresponding to this key. It also only allowed for one match value to be passed, whereas for many objects we need support for multiple match keys to guarantee uniqueness.
    - `get_matching_objects()` extracts the keys *and* values together based on the type-specific match key definitions in `mytardis_client` (see below). It also attempts to do identifier-based matching first, as `get_objects()` did, but does so by extracting the identifier from the object data, instead of trying the input string as an identifier and then a name.
    - `get_objects_by_identifier()` makes it more explicit when we just want to match by identifier.
    - Add `get_uris_by_identifier()` to make it clear when we are doing identifier-based matching.
- Overseer tests changes:
  - Update the tests to use the new methods in Overseer
  - Reinstate some `xfail`'d tests, after fixing a small fixture issue
- Move `resource_uri_to_id` from Overseer to `mytardis_client` as a free function, as this logic is closely coupled to how MyTardis constructs URIs.
- Separate information about MyTardis types (e.g. match fields) from endpoint information
  - Add a new module `src.mytardis_client.types` to store the type info
  - Add a new module `src.mytardis_client.endpoints` to store the endpoint info

JLoveUOA commented on src/overseers/overseer.py:
```diff
@@ -66,33 +88,55 @@ def mytardis_setup(self) -> IntrospectionConfig:
-        uri_sep: str = "/"
-        return int(urlparse(uri).path.rstrip(uri_sep).split(uri_sep).pop())
+        # pylint: disable=unsupported-membership-test
+        if object_type not in self.mytardis_setup.objects_with_ids:
```
Is this a duplication of type_supports_identifiers() above?

andrew-uoa replied:
Yeah, you're right. I've now replaced `type_supports_identifiers()` with a method `check_identifiers_enabled_for_type()`, which basically asserts both that identifiers are enabled generally, and that the specific type has identifiers enabled, and tried to replace most of the checking with that, except in `generate_object_matchers()` where we check if an object has identifiers enabled, but it's not error if they aren't.

JLoveUOA commented on src/overseers/overseer.py:
```diff
@@ -66,33 +88,55 @@ def mytardis_setup(self) -> IntrospectionConfig:
+            for identifier in identifiers:
+                yield {"identifier": identifier}
+
+        # NOTE: if there are no identifiers, we'll yield nothing - is that an error case?
```
It doesn't seem to cause any issues downstream and identifiers are optional so it feels like the intended outcome to generate nothing if iterating over an missing list.

JLoveUOA replied:
Or did you mean to add an error elsewhere if nothing can be found based on the uids and the identifiers?

andrew-uoa replied:
Good catch, I agree it's probably the right outcome so I removed the comment

JLoveUOA commented on src/overseers/overseer.py:
```diff
@@ -149,72 +195,46 @@ def get_object_by_uri(
 
-        Returns:
-            A list of objects from the search request made.
+        if not self.type_supports_identifiers(object_type):
```
Another duplicate of type_supports_identifiers()?

andrew-uoa replied:
Yeah thanks, see my reply above

JLoveUOA commented:
Looks really good. Only some minor notes, we talked through most of the rationale and it looks like everything is a lot tidier with these changes.

JLoveUOA commented:
it doesn't look like get_objects_by_identifier() is used anywhere? The current norm appears to be getting Uris from identifiers then retrieving objects from those uris. 
Always going through the uri  is nice for debugging. but is there a situation where we'd want to use get_objects_by_identifier() because it's a bit more efficient to get directly to an object from an identifier?

andrew-uoa commented:
> it doesn't look like get_objects_by_identifier() is used anywhere? The current norm appears to be getting Uris from identifiers then retrieving objects from those uris. Always going through the uri is nice for debugging. but is there a situation where we'd want to use get_objects_by_identifier() because it's a bit more efficient to get directly to an object from an identifier?

Thanks for the review, that's very helpful.

- Good point about `get_objects_by_identifier()`, I've removed that now.
- I couldn't find any place where we retrieve objects by URI, so I'm a bit confused - maybe I missed something? Your comment made me check and I realised we don't actually use `get_objects_by_uri()` anywhere, so I deleted it.
- I've made some changes to reduce the duplication (see replies to your comments) - let me know what you think

JLoveUOA commented on src/overseers/overseer.py:
```diff
@@ -66,33 +85,53 @@ def mytardis_setup(self) -> IntrospectionConfig:
 
-        Args:
-            uri: str - the URI from MyTardis
+        assert (
```
Nice that resolves a lot of the duplications. Bandit complains about raw Asserts due to them not working when compiled as optimized tho, so worth changing to explicitly raising errors (poteintally just to shut bandit up on future PRs).
